deployment/WFCORE-3522_Deployment content validation.adoc

```diff
@@ -0,0 +1,72 @@
+= Title
+:author:            David Lloyd
+:email:             david.lloyd@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          deployment
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+We need to move to child-first class and resource resolution in deployments.
+The first step must be validation of deployment contents and dependencies.
+This entails detecting deployment artifacts which have packages or resources which conflict with container packages or resources.
+
+Once this proposal is implemented, we can move to changing dependency order, which will come in a follow-up proposal.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-3522[WFCORE-3522]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/SWARM-1691[SWARM-1691]
+* https://issues.jboss.org/browse/WFLY-9828[WFLY-9828]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mjurc@redhat.com[Michal Jurc]
+
+=== Affected Projects or Components
+
+* WildFly Core
+* All deployment-related subsystems
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* The deployer must provide a way for subsystems to register packages and resources that are indicative of invalid deployment content.
+This likely entails introducing a multi-valued capability which is consumed by the deployment system.
+** The deployer must correlate this information in such a way that if the subsystem is disabled for a given deployment via ``exclude-subsystem`` in ``jboss-deployment-structure.xml``, the deployment can proceed without error.
+** There may be APIs that must always be forbidden, even when the ``exclude-subsystem`` flag is set.
+In these cases, the subsystem must have a way to indicate this situation.
+* The deployer must ignore invalid content.
```

Comment:
What about case when deployment contains enriched API or overriding content with additional dependencies that the 'valid' deployment content depends on?

Reply:
Since packages cannot be split, it is not allowed for deployments to supplement container-provided APIs with additional classes.  Additional packages which are not part of container-provided APIs are allowed.

Container classes cannot link to deployment classes, but they can load them (via ServiceLoader for example).  Some APIs (like JAXP or JSONP) support loading different per-deployment implementations.  Some (like JSF) support complete replacement of the API, but as stated in the requirements, users must exclude the subsystem to do so.  Some container provided APIs cannot be implemented or replaced by deployments, like JTA.

I've been working on [a spreadsheet](https://docs.google.com/spreadsheets/d/1sH0L8XLulQrMzXd8-ZWnxbej9xsD6W-NKzf9BJ7xB-c/edit#gid=0) which lists all of our container-available APIs and their level of replacement/implementation support.  It is not yet complete but it starting to give a picture.

Reply:
Thanks for the summary!

With regards to subsystems that do not allow to be disabled/overriden from deployments, is it planned to ensure that all the subsystems will allow deployments to disable/override them in future iterations of this feature? I kind of agree with your previous assessment that this feature would be way more useful and usable if all of the subsystems supported it.

Reply:
@mjurc I don't yet know if it's possible for every subsystem to support disabling/excluding itself from deployments.  Every subsystem that can possibly support doing so should be made to allow doing so.